x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/settings/TransportReloadRemoteClusterCredentialsAction.java

```diff
@@ -39,18 +45,38 @@ public class TransportReloadRemoteClusterCredentialsAction extends TransportActi
     ActionResponse.Empty> {
 
     private final RemoteClusterService remoteClusterService;
+    private final ClusterService clusterService;
 
     @Inject
-    public TransportReloadRemoteClusterCredentialsAction(TransportService transportService, ActionFilters actionFilters) {
+    public TransportReloadRemoteClusterCredentialsAction(
+        TransportService transportService,
+        ClusterService clusterService,
+        ActionFilters actionFilters
+    ) {
         super(ActionTypes.RELOAD_REMOTE_CLUSTER_CREDENTIALS_ACTION.name(), actionFilters, transportService.getTaskManager());
         this.remoteClusterService = transportService.getRemoteClusterService();
+        this.clusterService = clusterService;
     }
 
     @Override
     protected void doExecute(Task task, Request request, ActionListener<ActionResponse.Empty> listener) {
-        // We avoid stashing and marking context as system to keep the action as minimal as possible (i.e., avoid copying context)
-        remoteClusterService.updateRemoteClusterCredentials(request.getSettings());
-        listener.onResponse(ActionResponse.Empty.INSTANCE);
+        assert Transports.assertNotTransportThread("Remote connection re-building is too much for a transport thread");
+        final ClusterState clusterState = clusterService.state();
+        final ClusterBlockException clusterBlockException = checkBlock(clusterState);
+        if (clusterBlockException != null) {
+            throw clusterBlockException;
+        }
+        // Use a supplier to ensure we resolve cluster settings inside a synchronized block, to prevent race conditions
+        final Supplier<Settings> settingsSupplier = () -> {
```

Comment:
> Is the race condition this guards against a thread reading the settings, waiting for the lock and when they acquire the lock, the settings have been updated?

Yup exactly. 

> we don't want to add a dependency to `ClusterState`

Yeah it felt unwarranted to expose this to the `RemoteClusterService` since it's concerned primarily with remote connection handling and the use of `ClusterState` here is very narrow and targeted for the exact method introduced.